library/std/src/sys/sgx/thread.rs

```diff
@@ -104,7 +104,11 @@ pub mod wait_notify {
 
 impl Thread {
     // unsafe: see thread::Builder::spawn_unchecked for safety requirements
-    pub unsafe fn new(_stack: usize, p: Box<dyn FnOnce()>) -> io::Result<Thread> {
+    pub unsafe fn new(
+        _stack: usize,
+        p: Box<dyn FnOnce()>,
+        _native_options: BuilderOptions,
```

Comment:
Why are you modifying all of these system implementations?

Reply:
It's to support the trait extension. We need to define extra fields on the builder struct, and we want to avoid adding 3ds specific cfg sections to that code. So we need to give all operating systems the chance to add new fields to the builder struct. And how they use the new fields is also platform specific, so that needs to happen here. Thus, the data is passed even if it's not used (it would be optimized out).

Reply:
This is a big problem though. We should keep ourselves on `unix` only, and that’s already much. Changing all of these implementations isn’t something we can just push.

Reply:
We should take this to a _completely pthread_ level. We could use a global Mutex which holds the data of the “latest” building thread. It sounds absolutely trash, but we wouldn’t have to worry about `std` changes breaking the system. (We should flush the values stored in the mutex at thread creation, and keep everything in `pthread-3ds`, complete of renaming the functions to the standard declarations).

Reply:
1. These aren't random changes. We are doing a new thing here by allowing target specific thread builder properties. We don't want to hard code this for the 3DS only, so we need to give all platforms the ability to do this too. The other platforms don't use it right now though, so nothing changes functionally except for the 3DS.
2. This is in _absolutely no way_ modifying the public interface of the other platforms. There is no public change. These changes are all in private implementation modules. Maybe this is what you are tripping up on. There is no functional change to the other platforms.

Reply:
I worded it wrongly, by "public interface", I meant the one `sys` modules provide to the `thread` module. I meant changing how each `sys` module is supposed to work, just to accomodate for a feature only we need.

Reply:
Is it a feature we only use, but you guys asked for a trait extension (which is probably the right approach). You can't write a trait extension that adds fields to a struct, so we need a refactoring like this which supports adding fields to the thread builder in a target specific way.

At this point, we still don't _need_ to modify the other platforms, but we'd be hard coding it so only the 3DS is allowed to have custom thread builder properties. Based on what I saw in the other platforms' code, they would benefit from having this functionality too. Additionally, the point of having an imp module is to treat all platforms the same at a high level. That's why there's stuff like the imp::guard module defined in all platforms' code even though only a few support it. To keep the interface between the high level thread module and the platform specific modules stable/consistent, we need to implement no-op builder option support in all platforms.

There's a review process for std PRs for a reason. If they have an alternative way of doing this and would rather us do it differently, they will tell us.

Reply:
Don't be afraid of changing the sys module slightly. The whole point of that module is to be an internal (not stable) way to abstract over the various supported platforms. We can change it without fear.

Reply:
I think @AzureMarker makes some good points here, and I'm not entirely opposed to a change set like this, but I do still have a few concerns. I haven't really contributed to `std` before, so I don't know too much about the process.

* Should we try to upstream smaller, less invasive changes (like most of the `#[cfg(...)]` etc.) before trying to add new features like this? It seems to me that maintainers might prefer to review things in smaller digestible chunks.
* Is this the kind of thing that an RFC would be required for? It's a fairly small feature, but it's still a feature that lives behind a feature gate. If so, I definitely think it would be preferable to separate it from the rest of our std changes since it would require a tracking issue, etc.

I'm slightly wary of building up too many big code changes in this repo and trying to get them all merged at once, and separating them out later will probably be painful. Other than trying to upstream sooner, I'm not sure we have a very good way to gauge how maintainers feel about stuff like this.

Reply:
For the RFC question, I think we're fine to just do a PR since the public change (adding the trait extension) is behind an unstable feature flag.

Reply:
We could also make a Zulip thread (or wherever std people hang out) to get their opinion on it.

Reply:
We got some good feedback on the Zulip thread. The action items are:
1. Make a PR (or multiple) with the changes we already have now.
2. Make the thread changes in a separate PR.

See #6 for followups on item 1.